### ISSUE TITLE
API endpoints with primary keys that use Morango's UUID field return 500s for invalid IDs rather than a 4xx code

### ISSUE THREAD

**Issue by rtibbles** [learningequality/kolibri#9013]
## Observed behavior
Navigating to a URL such as `/api/content/contentnode/invalid/` will result in 500 errors due to the badly formed UUID.

## Errors and logs

```
ValueError: badly formed hexadecimal UUID string
  File "django/core/handlers/exception.py", line 41, in inner
    response = get_response(request)
  File "django/core/handlers/base.py", line 187, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "django/core/handlers/base.py", line 185, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "rest_framework/viewsets.py", line 116, in view
    return self.dispatch(request, *args, **kwargs)
  File "django/utils/decorators.py", line 67, in _wrapper
    return bound_func(*args, **kwargs)
  File "kolibri/core/auth/middleware.py", line 119, in wrapper_func
    return view(*args, **kwargs)
  File "kolibri/core/content/api.py", line 91, in wrapper_func
    response = some_func(*args, **kwargs)
  File "django/utils/decorators.py", line 63, in bound_func
    return func.__get__(self, type(self))(*args2, **kwargs2)
  File "rest_framework/views.py", line 495, in dispatch
    response = self.handle_exception(exc)
  File "rest_framework/views.py", line 455, in handle_exception
    self.raise_uncaught_exception(exc)
  File "rest_framework/views.py", line 492, in dispatch
    response = handler(request, *args, **kwargs)
  File "kolibri/core/api.py", line 346, in retrieve
    return Response(self.serialize_object())
  File "kolibri/core/api.py", line 265, in serialize_object
    return self.serialize(queryset.filter(**filter_kwargs))[0]
  File "kolibri/core/api.py", line 258, in serialize
    list(map(self._map_fields, values_queryset or [])), queryset
  File "django/db/models/query.py", line 254, in __bool__
    self._fetch_all()
  File "django/db/models/query.py", line 1121, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "django/db/models/query.py", line 106, in __iter__
    for row in compiler.results_iter(chunked_fetch=self.chunked_fetch):
  File "django/db/models/sql/compiler.py", line 841, in results_iter
    results = self.execute_sql(MULTI, chunked_fetch=chunked_fetch)
  File "django/db/models/sql/compiler.py", line 876, in execute_sql
    sql, params = self.as_sql()
  File "django/db/models/sql/compiler.py", line 441, in as_sql
    where, w_params = self.compile(self.where) if self.where is not None else ("", [])
  File "django/db/models/sql/compiler.py", line 373, in compile
    sql, params = node.as_sql(self, self.connection)
  File "django/db/models/sql/where.py", line 79, in as_sql
    sql, params = compiler.compile(child)
  File "django/db/models/sql/compiler.py", line 373, in compile
    sql, params = node.as_sql(self, self.connection)
  File "django/db/models/lookups.py", line 170, in as_sql
    rhs_sql, rhs_params = self.process_rhs(compiler, connection)
  File "django/db/models/lookups.py", line 103, in process_rhs
    return self.get_db_prep_lookup(value, connection)
  File "django/db/models/lookups.py", line 196, in get_db_prep_lookup
    [get_db_prep_value(value, connection, prepared=True)]
  File "morango/models/fields/uuids.py", line 39, in get_db_prep_value
    value = uuid.UUID(value)
  File "uuid.py", line 140, in __init__
    raise ValueError('badly formed hexadecimal UUID string')
```

## Expected behavior
Badly formed UUIDs should be properly handled and result in either a 404 or a 400. 404 would be preferable.

Unclear whether this should be fixed in the UUID field implementation in Morango, or just where it is used in Kolibri.

## User-facing consequences
Unhandled server errors are bad, mkay?

**Comment by sakshampathak1508** [comment]
Hey i would like to work on this issue.

**Comment by rtibbles** [comment]
Hi @sakshampathak1508 - that would be great!

The first thing to work out is what kind of error Django usually raises when an invalid field value is passed into a query. One quick way to check this would be to try passing a string in a query for a model type that has an integer primary key.

I don't know off the top of my head what error that should return - if it returns a `ValueError` then that means the current UUIDField implementation that we are using from our internal Morango library is behaving the same way as Django would, and so it's just a matter of handling this properly in our API endpoints.

If Django does something different, then we might need to modify Morango to behave more in line with Django.

**Comment by sakshampathak1508** [comment]
hello @rtibbles  i think in order to check the api endpoints the path would be something like `kolibri/core/content/api.py`
if i am correct please do tell me in which function we are calling it or rather returning the serialized data 

addition can't we use self.get_object_or_404 in this case or uuid has some role in it?

**Comment by rtibbles** [comment]
Sorry, I've just checked the traceback above - it seems the issue is coming from our custom implementation in `kolibri/core/api.py` that is only catching `IndexError` when trying to retrieve the object:

https://github.com/learningequality/kolibri/blob/develop/kolibri/core/api.py#L266

Instead we should ensure that we catch `ValueError` here as well, and maybe the other errors that are caught by `get_object_or_404`: https://github.com/encode/django-rest-framework/blob/3.9.x/rest_framework/generics.py#L22

**Comment by sakshampathak1508** [comment]
![invalid-path](https://user-images.githubusercontent.com/59732968/159155750-8d17b3fe-a4f5-4668-8e70-2b01865107ca.jpeg)
so i tested it by replacing the `except IndexError` to `except Exception` and it will do the trick and drf returning standard http 404 not found
should i raise a PR for that?

**Comment by rtibbles** [comment]
> `except IndexError` to except `Exception`

It's generally bad practice to catch `Exception` as that will catch _all_ errors, better to only catch the ones that we know are indicative of the object not existing: `except (IndexError, ValueError, TypeError):`

Would also be good to add a regression test for the fix within the PR.

**Comment by sakshampathak1508** [comment]
Actually i don't really know how to add tests its a section where i need to study but for the bug i made the changes and raising a PR if that's okay

**Comment by rtibbles** [comment]
Fixed in #9217
